src/ucp/wireup/wireup_ep.c

```diff
@@ -579,6 +579,243 @@ ssize_t ucp_wireup_ep_sockaddr_fill_private_data(void *arg, const char *dev_name
     return status;
 }
 
+static ucs_status_t
+ucp_wireup_sockaddr_select_tl_iface(ucp_worker_h worker, const char *dev_name,
+                                    ucp_worker_iface_t **wiface_p,
+                                    ucp_rsc_index_t *cm_idx_p)
+{
+    const size_t    dev_name_length = strlen(dev_name);
+    ucp_rsc_index_t num_tls;
+    ucp_rsc_index_t tl_rsc_idx, cm_idx, iface_idx, cmpt_idx;
+
+    num_tls = worker->context->num_tls;
+    for (tl_rsc_idx = 0; tl_rsc_idx < num_tls; ++tl_rsc_idx) {
+        if (strncmp(worker->context->tl_rscs[tl_rsc_idx].tl_rsc.dev_name,
+                    dev_name, dev_name_length)) {
+            continue;
+        }
+
+        for (iface_idx = 0; iface_idx < worker->num_ifaces; ++iface_idx) {
+            if (worker->ifaces[iface_idx].rsc_index != tl_rsc_idx) {
+                continue;
+            }
+
+            for (cm_idx = 0; cm_idx < ucp_worker_num_cm_cmpts(worker); ++cm_idx) {
+                cmpt_idx = ucp_context_rsc_idx_to_cmpt_idx(worker->context,
+                                                           tl_rsc_idx);
+                /*
+                 * FIXME: rdmacm CM implementation should be moved to IB
+                 *        component, then component indexes must match.
+                 *        This check is not critical until we have only 1 CM.
+                 */
+                if (1 || (cmpt_idx == worker->cms[cm_idx].cmpt_idx)) {
+                    *wiface_p = &worker->ifaces[iface_idx];
+                    *cm_idx_p = cm_idx;
+                    /* TODO: select the best one */
+                    return UCS_OK;
+                }
+            }
+        }
+    }
+
+    return UCS_ERR_NO_DEVICE;
+}
+
+static void
+ucp_wireup_sockaddr_set_tl_lane(ucp_ep_h ucp_ep, uct_ep_h tl_ep,
+                                ucp_rsc_index_t tl_rsc_idx)
+{
+    ucp_ep_config_key_t key     = ucp_ep_config(ucp_ep)->key;
+    ucp_lane_index_t wireup_idx = key.wireup_lane;
+    ucp_wireup_ep_t *wireup_ep  = (ucp_wireup_ep_t *)ucp_ep->uct_eps[wireup_idx];
+    ucs_status_t status;
+
+    key.lanes[wireup_idx].rsc_index = tl_rsc_idx;
+    status = ucp_worker_get_ep_config(ucp_ep->worker, &key, 0,
+                                      &ucp_ep->cfg_index);
+    ucs_assert_always(status == UCS_OK);
+    ucp_wireup_ep_set_next_ep(&wireup_ep->super.super, tl_ep);
+}
+
+ssize_t ucp_wireup_sockaddr_cm_priv_pack_cb(void *arg, const char *dev_name,
+                                            void *priv_data)
+{
+    ucp_wireup_client_data_t *client_data = priv_data;
```

Comment:
since this function is called both on the server and client side to pack the data, please rename the struct and param so that they won't have 'client' in the name.